### PR TITLE
HTTP+JSON serialization sink

### PR DESCRIPTION
**Pull request by theatrus**
In order to reduce pull request size and interdependence,
here is an incremental step for a stdout JSON output sink. The stdout
writes will of course be replaced with a lifoq push for libcurl posting.

The current sink infrastructure is a bit hard to unit test - once the
HTTP plumbing is in place, there is an integration test for it.

### CODE REVIEW

File: src/sink.c [hunk 1]
```diff
@@ -3,16 +3,26 @@
...
-            }
+            actual_sink->next = *sinks;
             *sinks = actual_sink;
-        } else {
+            break;
+        }
+        case SINK_TYPE_HTTP:
+        {
+            sink* actual_sink = init_http_sink((sink_config_http*)sc, config);
+            actual_sink->next = *sinks;
```
Comment (smparkes) [inline src/sink.c]: Could factor out the list plumbing.

File: src/sink_http.c [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+
+
+typedef struct http_sink {
+    sink sink;
+} http_sink;
+
+/*
+ * Data from the metrics_iter callback */
+struct cb_info {
```
Comment (smparkes) [inline src/sink_http.c:31]: typedef for consistency?

File: src/sink_http.c [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+     * Scary looking macro to apply suffixes to a string, and then
+     * insert them into a json object with a given value. Needs "suffixed"
+     * in scope, with a "base_len" telling it how mmany chars the string is
+     */
+#define SUFFIX_ADD(suf, val)                                            \
+    do {                                                                \
+        suffixed[base_len - 1] = '\0';                                  \
+        strcat(suffixed, suf);                                          \
+        json_object_set_new(obj, suffixed, val);                        \
+    } while(0);
```
Comment (smparkes) [inline src/sink_http.c]: No ;?

File: src/sink_http.c [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    case KEY_VAL:
+        json_object_set_new(obj, full_name, json_real(*(double*)value));
+        break;
+    case GAUGE:
+        json_object_set_new(obj, full_name, json_real(*(double*)value));
+        break;
+    case COUNTER:
+    {
+        if (config->extended_counters) {
+            char suffixed[base_len + 8];
```
Comment (smparkes) [inline src/sink_http.c]: factor our magic 8 and 20 below to constants with comments on the max value (esp. for timers below where it's more complicated)?

File: src/sink_http.c [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    return 0;
+}
+
+static int json_cb(const char* buf, size_t size, void* d) {
+    strbuf* sbuf = (strbuf*)d;
+    strbuf_cat(sbuf, buf, size);
+    return 0;
+}
+
+static int serialize_metrics(http_sink* sink, metrics* m, void* data) {
```
Comment (smparkes) [inline src/sink_http.c]: Be nice to have tests for this.

File: src/sink.c (hunk 1 above)
Comment (smparkes) [inline src/sink.c:21]: factor list plumbing?

File: src/sink_http.c (hunk 2 above)
Comment (theatrus) [inline src/sink_http.c:31]: Trying to leave typedefs for exported types, so I removed the top typedef.

File: src/sink_http.c (hunk 4 above)
Comment (theatrus) [inline src/sink_http.c]: Now converted to snprintfs with explicit terminators

File: src/sink.c (hunk 1 above)
Comment (theatrus) [inline src/sink.c:21]: Yeah, I'll look at some macros for single LL intrusive operations. Its mostly straightforward. I'll open an issue for this one as a number of places could benefit.

File: src/sink_http.c (hunk 5 above)
Comment (theatrus) [inline src/sink_http.c]: the whole serialize workflow? Yeah...

A bit awkward with how the test suite is run right now, where testing is only possible on public interfaces. 

I'm going to add a second constructor interface so I don't leak all of the internal state to a test and check serialization output.

**Pull request merged**